src/diffusers/models/unet_2d_condition.py

```diff
@@ -1074,8 +1074,8 @@ def forward(
                     f"{self.__class__} has the config param `encoder_hid_dim_type` set to 'ip_image_proj' which requires the keyword argument `image_embeds` to be passed in  `added_conditions`"
                 )
             image_embeds = added_cond_kwargs.get("image_embeds")
-            image_embeds = self.encoder_hid_proj(image_embeds).to(encoder_hidden_states.dtype)
-            encoder_hidden_states = torch.cat([encoder_hidden_states, image_embeds], dim=1)
+            image_embeds = self.encoder_hid_proj(image_embeds)
+            encoder_hidden_states = (encoder_hidden_states, image_embeds)
```

Comment:
This is backwards breaking - I don't think we should do this change here (cc @yiyixuxu - do we really have to do this? Can't we adapt the attention processor to handle tensors just like before?)

Reply:
Previously, we only needed support 1 ip-adatper and 1 image. so it is easy to split out the image feature from the `encoder_hidden_states` tensor inside attention processor based on `self.num_tokens`
            
            end_pos = encoder_hidden_states.shape[1] - self.num_tokens
            encoder_hidden_states, ip_hidden_states = (
                        encoder_hidden_states[:, :end_pos, :],
                        encoder_hidden_states[:, end_pos:, :],
                    )

In this PR, we are trying to support multiple ip-adapters, and each ip-adapter can accept a list of image inputs with an arbitrary length. so the image feature for each ip-adapter depends on the adapter's `num_tokens` config and the number of images. In order to handle this inside attention processor, we will need to pass down the additional information about the number of images for each adapter during run time 
e.g. `num_image_inputs = [[1], [10]]`

and inside the attention processor, we can figure out the dimension of image feature for each ip-adapter

```python
 
            num_tokens = [n_image * n_token for n_image, n_token in zip(num_Image_inputs, self.num_tokens)]
            end_pos = encoder_hidden_states.shape[1] - sum(num_tokens)
            encoder_hidden_states, ip_hidden_states = (
                        encoder_hidden_states[:, :end_pos, :],
                        encoder_hidden_states[:, end_pos:, :],
                    )
           
          for n_tokens in num_tokens:
                  end_pos +=  n_tokens
                  current_ip_hidden_states, ip_hidden_states = (
                       ip_hidden_states[:,:end_pos,:], (ip_hidden_states[:,end_pos:,:],
                  )
                 ....
          
```  

we can do that by adding a new argument to the attention processor and then pass the variable down via `cross_attention_kwargs`. However, more importantly,  I don't understand why it would be backward breaking?  this code here only affects IP-Adapter. and I've tested it would works with previous API when you only pass one image and one ip-adapter.

Reply:
It might be too much of an edge case, but we allow users to define custom attention processors and then set them via:

`unet.set_attn_processor(...)` 

If someone defined a custom attention processor for IP adapter that expects a single tensor as input for `encoder_hidden_states`, it would be broken. But maybe that's also too much of an edge case. 

Ok to leave then as is (maybe you could throw a nice error message in the attention processor if `encoder_hidden_states` is not a tuple that says that a tuple of tensors is now expected)

Reply:
ok, I got it now! I will keep this use case in mind in the future!

In this case, I think I would just throw a nice error message here - I could rewrite to code to make it backward compatible. However, it will make the code a lot more complex and I think it is not worth it here